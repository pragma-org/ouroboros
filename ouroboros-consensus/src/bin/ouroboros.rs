use std::{thread, time::Duration};

use ouroboros_consensus::{message::Message, network::Network};
use ouroboros_simulation::{network::random_topology, parameters::Parameters};
use tracing_subscriber::{fmt::format::FmtSpan, prelude::*, Registry};

pub fn main() {
    let subscriber = Registry::default();
    let json_log = tracing_subscriber::fmt::layer()
        .with_span_events(FmtSpan::CLOSE)
        .with_target(false)
        .with_level(false)
        .json();
    let subscriber = subscriber.with(json_log);

    tracing::subscriber::set_global_default(subscriber).expect("Unable to set global subscriber");

    let parameters = Parameters::default();
    let topology = random_topology(&mut rand::thread_rng(), &parameters);
    let network = Network::new(&topology, &parameters);
    let mut handle = network.start();
    for i in 0..1000 {
        thread::sleep(Duration::from_millis(10));
        handle.broadcast(Message::NextSlot(i));
    }

    handle.stop();

    for node_id in 1..10 {
        let chain = handle.get_preferred_chain(node_id);
        println!(
            "{{\"{}\": {}}}",
            node_id,
            serde_json::to_string(&chain).unwrap()
        );
    }
}
