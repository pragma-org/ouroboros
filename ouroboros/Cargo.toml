[package]
name = "ouroboros"
version = "0.1.0"
edition = "2021"

[dependencies]
hex = "0.4.3"
mockall = "0.13"
pallas-codec = { git = "https://github.com/txpipe/pallas", rev = "252714d58a93d58bf2cd7ada1156e3ace10a8298" }
pallas-crypto = { git = "https://github.com/txpipe/pallas", rev = "252714d58a93d58bf2cd7ada1156e3ace10a8298" }
thiserror = "1.0"

[dev-dependencies]
insta = { version = "1.40.0", features = ["yaml"] }