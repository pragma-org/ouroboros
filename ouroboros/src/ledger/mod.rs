use mockall::automock;
use pallas_codec::utils::Bytes;
use pallas_crypto::hash::{Hash, Hasher};
use thiserror::Error;

#[derive(Error, Debug)]
pub enum Error {
    #[error("PoolId not found")]
    PoolIdNotFound,
}

pub type PoolId = Hash<28>;

/// The sigma value of a pool. This is a rational number that represents the total value of the
/// delegated stake in the pool over the total value of the active stake in the network. This value
/// is tracked in the ledger state and recorded as a snapshot value at each epoch.
#[derive(Debug, PartialEq, Eq, Clone)]
pub struct PoolSigma {
    pub numerator: u64,
    pub denominator: u64,
}

/// The pool info trait provides a lookup mechanism for pool data. This is sourced from the ledger
#[automock]
pub trait PoolInfo: Send + Sync {
    /// Performs a lookup of a pool_id to its sigma value. This usually represents a different set of
    /// sigma snapshot data depending on whether we need to look up the pool_id in the current epoch
    /// or in the future.
    fn sigma(&self, pool_id: &PoolId) -> Result<PoolSigma, Error>;

    /// Hashes the vrf vkey of a pool.
    fn vrf_vkey_hash(&self, pool_id: &PoolId) -> Result<Hash<32>, Error>;
}

/// The node's cold vkey is hashed with blake2b224 to create the pool id
pub fn issuer_vkey_to_pool_id(issuer_vkey: &Bytes) -> PoolId {
    Hasher::<224>::hash(issuer_vkey)
}

#[cfg(test)]
mod tests {
    use crate::ledger::issuer_vkey_to_pool_id;
    use pallas_codec::utils::Bytes;

    #[test]
    fn test_issuer_vkey_to_pool_id() {
        let test_vector = vec![(
            "cad3c900ca6baee9e65bf61073d900bfbca458eeca6d0b9f9931f5b1017a8cd6",
            "00beef0a9be2f6d897ed24a613cf547bb20cd282a04edfc53d477114",
        )];
        insta::assert_yaml_snapshot!(test_vector);

        for (issuer_vkey_str, expected_pool_id_str) in test_vector {
            let issuer_vkey: Bytes = issuer_vkey_str.parse().unwrap();
            let pool_id = issuer_vkey_to_pool_id(&issuer_vkey);
            assert_eq!(pool_id.to_string(), expected_pool_id_str);
        }
    }
}
