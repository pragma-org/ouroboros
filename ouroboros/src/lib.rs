pub mod ledger;
pub mod validator;
