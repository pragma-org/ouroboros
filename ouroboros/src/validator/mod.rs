/// Generic trait for validating any type of data. Designed to be used across threads so validations
/// can be done in parallel. Validators should handle all error cases internally and simply return
/// a boolean indicating if the validation was successful.
pub trait Validator: Send + Sync {
    fn validate(&self) -> bool;
}
