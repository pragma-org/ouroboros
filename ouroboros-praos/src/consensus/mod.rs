use ouroboros::ledger::{issuer_vkey_to_pool_id, PoolId, PoolInfo};
use ouroboros::validator::Validator;
use pallas_crypto::hash::{Hash, Hasher};
use pallas_crypto::vrf::{
    VrfProof, VrfProofBytes, VrfProofHashBytes, VrfPublicKey, VrfPublicKeyBytes,
};
use pallas_math::math::{ExpOrdering, FixedDecimal, FixedPrecision};
use pallas_primitives::babbage;
use pallas_primitives::babbage::{derive_tagged_vrf_output, VrfDerivation};
use std::ops::Deref;
use std::sync::LazyLock;
use tracing::{error, span, trace, warn};

/// The certified natural max value represents 2^256 in praos consensus
static CERTIFIED_NATURAL_MAX: LazyLock<FixedDecimal> = LazyLock::new(|| {
    FixedDecimal::from_str(
        "1157920892373161954235709850086879078532699846656405640394575840079131296399360000000000000000000000000000000000",
        34,
    )
        .expect("Infallible")
});

/// Validator for a block using praos consensus.
pub struct BlockValidator<'b> {
    header: &'b babbage::Header,
    pool_info: &'b dyn PoolInfo,
    epoch_nonce: &'b Hash<32>,
    // c is the ln(1-active_slots_coeff). Usually ln(1-0.05)
    c: &'b FixedDecimal,
}

impl<'b> BlockValidator<'b> {
    pub fn new(
        header: &'b babbage::Header,
        pool_info: &'b dyn PoolInfo,
        epoch_nonce: &'b Hash<32>,
        c: &'b FixedDecimal,
    ) -> Self {
        Self {
            header,
            pool_info,
            epoch_nonce,
            c,
        }
    }

    fn validate_babbage_compatible(&self) -> bool {
        let span = span!(tracing::Level::TRACE, "validate_babbage_compatible");
        let _enter = span.enter();

        // Grab all the values we need to validate the block
        let issuer_vkey = &self.header.header_body.issuer_vkey;
        let pool_id: PoolId = issuer_vkey_to_pool_id(issuer_vkey);
        let vrf_vkey: VrfPublicKeyBytes = match (&self.header.header_body.vrf_vkey).try_into() {
            Ok(vrf_vkey) => vrf_vkey,
            Err(error) => {
                error!("Could not convert vrf_vkey: {}", error);
                return false;
            }
        };
        if !self.ledger_matches_block_vrf_key_hash(&pool_id, &vrf_vkey) {
            // Fail fast if the vrf key hash in the block does not match the ledger
            return false;
        }
        let sigma: FixedDecimal = match self.pool_info.sigma(&pool_id) {
            Ok(sigma) => {
                FixedDecimal::from(sigma.numerator) / FixedDecimal::from(sigma.denominator)
            }
            Err(error) => {
                warn!("{:?} - {:?}", error, pool_id);
                return false;
            }
        };
        let absolute_slot = self.header.header_body.slot;

        // Get the leader VRF output hash from the block vrf result
        let leader_vrf_output = &self.header.header_body.leader_vrf_output();

        let block_vrf_proof_hash: VrfProofHashBytes =
            match (&self.header.header_body.vrf_result.0).try_into() {
                Ok(block_vrf_proof_hash) => block_vrf_proof_hash,
                Err(error) => {
                    error!("Could not convert block vrf proof hash: {}", error);
                    return false;
                }
            };
        let block_vrf_proof: VrfProofBytes =
            match (&self.header.header_body.vrf_result.1).try_into() {
                Ok(block_vrf_proof) => block_vrf_proof,
                Err(error) => {
                    error!("Could not convert block vrf proof: {}", error);
                    return false;
                }
            };
        let kes_signature = self.header.body_signature.as_slice();

        trace!("pool_id: {}", pool_id);
        trace!("block vrf_vkey: {}", hex::encode(vrf_vkey));
        trace!("sigma: {}", sigma);
        trace!("absolute_slot: {}", absolute_slot);
        trace!("leader_vrf_output: {}", hex::encode(leader_vrf_output));
        trace!(
            "block_vrf_proof_hash: {}",
            hex::encode(block_vrf_proof_hash.as_slice())
        );
        trace!(
            "block_vrf_proof: {}",
            hex::encode(block_vrf_proof.as_slice())
        );
        trace!("kes_signature: {}", hex::encode(kes_signature));

        // Calculate the VRF input seed so we can verify the VRF output against it.
        let vrf_input_seed = self.mk_vrf_input(absolute_slot, self.epoch_nonce.as_ref());
        trace!("vrf_input_seed: {}", vrf_input_seed);

        // Verify the VRF proof
        let vrf_proof = VrfProof::from(&block_vrf_proof);
        let vrf_vkey = VrfPublicKey::from(&vrf_vkey);
        match vrf_proof.verify(&vrf_vkey, vrf_input_seed.as_ref()) {
            Ok(proof_hash) => {
                if proof_hash.as_slice() != block_vrf_proof_hash.as_slice() {
                    error!("VRF proof hash mismatch");
                    false
                } else {
                    // The proof was valid. Make sure that our leader_vrf_output matches what was in the block
                    trace!("certified_proof_hash: {}", hex::encode(proof_hash));
                    let calculated_leader_vrf_output =
                        derive_tagged_vrf_output(proof_hash.as_slice(), VrfDerivation::Leader);
                    if calculated_leader_vrf_output.as_slice() != leader_vrf_output.as_slice() {
                        error!(
                            "Leader VRF output hash mismatch. was: {}, expected: {}",
                            hex::encode(calculated_leader_vrf_output),
                            hex::encode(leader_vrf_output)
                        );
                        false
                    } else {
                        // The leader VRF output hash matches what was in the block
                        // Now we need to check if the pool had enough sigma stake to produce this block
                        if self.pool_meets_delegation_threshold(
                            &sigma,
                            absolute_slot,
                            leader_vrf_output.as_slice(),
                        ) {
                            // TODO: Validate the KES signature
                            true
                        } else {
                            false
                        }
                    }
                }
            }
            Err(error) => {
                error!("Could not verify block vrf: {}", error);
                false
            }
        }
    }

    /// Verify that the pool meets the delegation threshold
    fn pool_meets_delegation_threshold(
        &self,
        sigma: &FixedDecimal,
        absolute_slot: u64,
        leader_vrf_output: &[u8],
    ) -> bool {
        let certified_leader_vrf: FixedDecimal = leader_vrf_output.into();
        let denominator = CERTIFIED_NATURAL_MAX.deref() - &certified_leader_vrf;
        let recip_q = CERTIFIED_NATURAL_MAX.deref() / &denominator;
        let x = -(sigma * self.c);

        trace!("certified_leader_vrf: {}", certified_leader_vrf);
        trace!("denominator: {}", denominator);
        trace!("recip_q: {}", recip_q);
        trace!("c: {}", self.c);
        trace!("x: {}", x);

        let ordering = x.exp_cmp(1000, 3, &recip_q);
        match ordering.estimation {
            ExpOrdering::LT => {
                trace!(
                    "Slot: {} - IS Leader: {} < {}",
                    absolute_slot,
                    recip_q,
                    ordering.approx
                );
                true
            }
            _ => {
                trace!(
                    "Slot: {} - NOT Leader: {} >= {}",
                    absolute_slot,
                    recip_q,
                    ordering.approx
                );
                false
            }
        }
    }

    /// Validate that the VRF key hash in the block matches the VRF key hash in the ledger
    fn ledger_matches_block_vrf_key_hash(
        &self,
        pool_id: &PoolId,
        vrf_vkey: &VrfPublicKeyBytes,
    ) -> bool {
        let vrf_vkey_hash: Hash<32> = Hasher::<256>::hash(vrf_vkey);
        trace!("block vrf_vkey_hash: {}", hex::encode(vrf_vkey_hash));
        let ledger_vrf_vkey_hash = match self.pool_info.vrf_vkey_hash(pool_id) {
            Ok(ledger_vrf_vkey_hash) => ledger_vrf_vkey_hash,
            Err(error) => {
                warn!("{:?} - {:?}", error, pool_id);
                return false;
            }
        };
        if vrf_vkey_hash != ledger_vrf_vkey_hash {
            error!(
                "VRF vkey hash in block ({}) does not match registered ledger vrf vkey hash ({})",
                hex::encode(vrf_vkey_hash),
                hex::encode(ledger_vrf_vkey_hash)
            );
            return false;
        }
        true
    }

    fn mk_vrf_input(&self, absolute_slot: u64, eta0: &[u8]) -> Hash<32> {
        trace!("mk_vrf_input() absolute_slot {}", absolute_slot);
        let mut hasher = Hasher::<256>::new();
        hasher.input(&absolute_slot.to_be_bytes());
        hasher.input(eta0);
        hasher.finalize()
    }
}

impl Validator for BlockValidator<'_> {
    fn validate(&self) -> bool {
        self.validate_babbage_compatible()
    }
}

#[cfg(test)]
mod tests {
    use crate::consensus::BlockValidator;
    use ctor::ctor;
    use mockall::predicate::eq;
    use ouroboros::ledger::{MockPoolInfo, PoolId, PoolSigma};
    use ouroboros::validator::Validator;
    use pallas_crypto::hash::Hash;
    use pallas_math::math::{FixedDecimal, FixedPrecision};
    use pallas_traverse::MultiEraHeader;

    #[ctor]
    fn init() {
        // set rust log level to TRACE
        // std::env::set_var("RUST_LOG", "ouroboros-praos=trace");

        // initialize tracing crate
        tracing_subscriber::fmt::init();
    }

    #[test]
    fn test_validate_conway_block() {
        let test_block = include_bytes!("../../tests/data/mainnet_blockheader_10817298.cbor");
        let test_block_hex = hex::encode(test_block);
        insta::assert_snapshot!(test_block_hex);
        let test_vector = vec![
            (
                "00beef0a9be2f6d897ed24a613cf547bb20cd282a04edfc53d477114",
                "c0d1f9b040d2f6fd7fc8775d24753d6db4b697429f11404a6178a0a4a005867b",
                "c7937fc47fecbe687891b3decd71e904d1e129598aa3852481d295eea3ea3ada",
                25626202470912_u64,
                22586623335121436_u64,
                true,
            ),
            (
                "00beef0a9be2f6d897ed24a613cf547bb20cd282a04edfc53d477114",
                "c0d1f9b040d2f6fd7fc8775d24753d6db4b697429f11404a6178a0a4a005867b",
                "c7937fc47fecbe687891b3decd71e904d1e129598aa3852481d295eea3ea3ada",
                6026202470912_u64,
                22586623335121436_u64,
                false,
            ),
        ];
        insta::assert_yaml_snapshot!(test_vector);

        for (pool_id_str, vrf_vkey_hash_str, epoch_nonce_str, numerator, denominator, expected) in
            test_vector
        {
            let pool_id: PoolId = pool_id_str.parse().unwrap();
            let vrf_vkey_hash: Hash<32> = vrf_vkey_hash_str.parse().unwrap();
            let epoch_nonce: Hash<32> = epoch_nonce_str.parse().unwrap();

            let active_slots_coeff: FixedDecimal =
                FixedDecimal::from(5u64) / FixedDecimal::from(100u64);
            let c = (FixedDecimal::from(1u64) - active_slots_coeff).ln();
            let conway_block_tag: u8 = 6;
            let multi_era_header =
                MultiEraHeader::decode(conway_block_tag, None, test_block).unwrap();
            let babbage_header = multi_era_header.as_babbage().expect("Infallible");
            assert_eq!(babbage_header.header_body.slot, 134402628u64);

            let mut pool_info = MockPoolInfo::new();
            pool_info
                .expect_sigma()
                .with(eq(pool_id))
                .returning(move |_| {
                    Ok(PoolSigma {
                        numerator,
                        denominator,
                    })
                });
            pool_info
                .expect_vrf_vkey_hash()
                .with(eq(pool_id))
                .returning(move |_| Ok(vrf_vkey_hash));

            let block_validator = BlockValidator::new(babbage_header, &pool_info, &epoch_nonce, &c);
            assert_eq!(block_validator.validate(), expected);
        }
    }
}
