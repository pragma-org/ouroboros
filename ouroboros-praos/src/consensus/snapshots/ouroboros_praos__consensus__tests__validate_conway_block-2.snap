---
source: ouroboros-praos/src/consensus/mod.rs
expression: test_vector
---
- - 00beef0a9be2f6d897ed24a613cf547bb20cd282a04edfc53d477114
  - c0d1f9b040d2f6fd7fc8775d24753d6db4b697429f11404a6178a0a4a005867b
  - c7937fc47fecbe687891b3decd71e904d1e129598aa3852481d295eea3ea3ada
  - 25626202470912
  - 22586623335121436
  - true
- - 00beef0a9be2f6d897ed24a613cf547bb20cd282a04edfc53d477114
  - c0d1f9b040d2f6fd7fc8775d24753d6db4b697429f11404a6178a0a4a005867b
  - c7937fc47fecbe687891b3decd71e904d1e129598aa3852481d295eea3ea3ada
  - 6026202470912
  - 22586623335121436
  - false
